vault/identity_store_test.go

```diff
@@ -1416,61 +1417,53 @@ func TestIdentityStoreInvalidate_TemporaryEntity(t *testing.T) {
 // the identity cleanup rename resolver to ensure that loading is deterministic
 // for both.
 func TestIdentityStoreLoadingIsDeterministic(t *testing.T) {
-	for i := 0; i < 50; i++ {
-		seedval := rand.Int63()
-		if os.Getenv("VAULT_TEST_IDENTITY_STORE_SEED") != "" {
-			var err error
-			seedval, err = strconv.ParseInt(os.Getenv("VAULT_TEST_IDENTITY_STORE_SEED"), 10, 64)
-			require.NoError(t, err)
-		}
-		seed := rand.New(rand.NewSource(seedval)) // Seed for deterministic test
-		defer t.Logf("Test generated with seed: %d", seedval)
+	seedval := rand.Int63()
+	if os.Getenv("VAULT_TEST_IDENTITY_STORE_SEED") != "" {
+		var err error
+		seedval, err = strconv.ParseInt(os.Getenv("VAULT_TEST_IDENTITY_STORE_SEED"), 10, 64)
+		require.NoError(t, err)
+	}
+	defer t.Logf("Test generated with seed: %d", seedval)
 
-		tests := []struct {
-			name  string
-			flags *determinismTestFlags
-		}{
-			{
-				name: "error-resolver-primary",
-				flags: &determinismTestFlags{
-					identityDeduplication: false,
-					secondary:             false,
-					seed:                  seed,
-				},
-			},
-			{
-				name: "identity-cleanup-primary",
-				flags: &determinismTestFlags{
-					identityDeduplication: true,
-					secondary:             false,
-					seed:                  seed,
-				},
-			},
+	tests := map[string]*determinismTestFlags{
+		"error-resolver-primary": {
+			identityDeduplication: false,
+			secondary:             false,
+		},
+		"identity-cleanup-primary": {
+			identityDeduplication: true,
+			secondary:             false,
+		},
+	}
 
-			{
-				name: "error-resolver-secondary",
-				flags: &determinismTestFlags{
-					identityDeduplication: false,
-					secondary:             true,
-					seed:                  seed,
-				},
-			},
-			{
-				name: "identity-cleanup-secondary",
-				flags: &determinismTestFlags{
-					identityDeduplication: true,
-					secondary:             true,
-					seed:                  seed,
-				},
-			},
+	// Hook to add cases that only differ in Enterprise
+	if entIdentityStoreDeterminismSupportsSecondary() {
+		tests["error-resolver-secondary"] = &determinismTestFlags{
+			identityDeduplication: false,
+			secondary:             true,
 		}
-
-		for _, test := range tests {
-			t.Run(t.Name()+"-"+test.name, func(t *testing.T) {
-				identityStoreLoadingIsDeterministic(t, test.flags)
-			})
+		tests["identity-cleanup-secondary"] = &determinismTestFlags{
+			identityDeduplication: true,
+			secondary:             true,
 		}
 	}
+
+	repeats := 50
+
+	for name, flags := range tests {
+		t.Run(t.Name()+"-"+name, func(t *testing.T) {
+			// Create a random source specific to this test case so every test case
+			// starts out from the identical random state given the same seed. We do
+			// want each iteration to explore different path though so we do it here
+			// not inside the test func.
+			seed := rand.New(rand.NewSource(seedval)) // Seed for deterministic test
+			flags.seed = seed
+
```

Comment:
Do you think we should keep this?

Reply:
I don't mind removing it, but it has helped us find problems earlier by running different seeds as part of the same test.

Reply:
No worries either way -- I don't insist! In my mind we could accomplish the same thing by using `untilfail`, but it may also be useful to have multiple test runs.

My only concerns are:
1. Test bloat on this runner.
2. Potential for no net-gain having this in CI, since we retry failures to weed out flakes, and that might mask a failing seed.

Reply:
I think it's useful to keep: this test only caught some legit non-determinism reliably after 20+ iterations. If we just loop once that kind of bug will just cause a 1-in-20 flake in CI and no one will ever bother looking at it. Having this in CI will keep us honest - this is effectively a "fuzz" test so more iterations == stronger test.